File: front-end/src/services/rulesService.js
import { db } from "../components/firebase.js";
import {
  collection,
  getDocs,
  addDoc,
  updateDoc,
  deleteDoc,
  doc,
  query,
  where,
  orderBy,
  getDoc,
} from "firebase/firestore";

// Import JSON data
import behaviorsData from "../components/rules/bieuhien.json";
import plusData from "../components/rules/diemcong.json";
import minusData from "../components/rules/diemtru.json";

class RulesService {
  constructor() {
    this.behaviorsData = behaviorsData;
    this.plusData = plusData;
    this.minusData = minusData;
  }

  // Calculate total behavior points
  calculateTotalBehaviorPoints() {
    let total = 0;
    Object.values(this.behaviorsData).forEach((category) => {
      Object.values(category.criteria).forEach((criterion) => {
        total += criterion.point;
      });
    });
    return total;
  }

  // Get all rules data
  getAllRulesData() {
    return {
      behaviors: this.behaviorsData,
      plus: this.plusData,
      minus: this.minusData,
    };
  }

  // Get rules by type
  getRulesByType(type) {
    switch (type) {
      case "behaviors":
        return this.behaviorsData;
      case "plus":
        return this.plusData;
      case "minus":
        return this.minusData;
      default:
        return {};
    }
  }

  // Generate next code for a rule
  generateNextCode(parentCode, type) {
    const data = this.getRulesByType(type);
    const category = Object.values(data).find((cat) =>
      Object.values(cat.criteria).some((crit) => crit.code === parentCode)
    );

    if (!category) return `${parentCode}.A1`;

    const criterion = Object.values(category.criteria).find(
      (crit) => crit.code === parentCode
    );
    if (!criterion) return `${parentCode}.A1`;

    const items =
      criterion.behaviors || criterion.plus || criterion.minus || {};
    const existingCodes = Object.keys(items);

    if (existingCodes.length === 0) {
      const suffix = type === "behaviors" ? "A" : type === "plus" ? "C" : "B";
      return `${parentCode}.${suffix}1`;
    }

    // Find the highest number
    const numbers = existingCodes.map((code) => {
      const match = code.match(new RegExp(`${parentCode}\\.([ABC])(\\d+)`));
      return match ? parseInt(match[2]) : 0;
    });

    const maxNumber = Math.max(...numbers);
    const suffix = type === "behaviors" ? "A" : type === "plus" ? "C" : "B";
    return `${parentCode}.${suffix}${maxNumber + 1}`;
  }

  // Get rules from Firebase
  async getRulesFromFirebase() {
    try {
      const rulesRef = collection(db, "rules_items");
      const q = query(rulesRef, orderBy("createdAt", "desc"));
      const querySnapshot = await getDocs(q);

      return querySnapshot.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
        createdAt: doc.data().createdAt?.toDate?.() || new Date(),
      }));
    } catch (error) {
      console.error("Error fetching rules from Firebase:", error);
      return [];
    }
  }

  // Get criteria from Firebase
  async getCriteriaFromFirebase() {
    try {
      const criteriaRef = collection(db, "criteria_info");
      const q = query(criteriaRef, orderBy("createdAt", "desc"));
      const querySnapshot = await getDocs(q);

      return querySnapshot.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
        createdAt: doc.data().createdAt?.toDate?.() || new Date(),
      }));
    } catch (error) {
      console.error("Error fetching criteria from Firebase:", error);
      return [];
    }
  }

  // Add criteria to Firebase
  async addCriteriaToFirebase(criteriaData) {
    try {
      const criteriaRef = collection(db, "criteria_info");
      const docRef = await addDoc(criteriaRef, {
        ...criteriaData,
        createdAt: new Date(),
      });
      return docRef.id;
    } catch (error) {
      console.error("Error adding criteria to Firebase:", error);
      throw error;
    }
  }

  // Add a new rule to Firebase
  async addRuleToFirebase(ruleData) {
    try {
      const rulesRef = collection(db, "rules_items");
      const docRef = await addDoc(rulesRef, {
        ...ruleData,
        createdAt: new Date(),
      });
      return docRef.id;
    } catch (error) {
      console.error("Error adding rule to Firebase:", error);
      throw error;
    }
  }

  // Update a rule in Firebase
  async updateRuleInFirebase(ruleId, ruleData) {
    try {
      const ruleRef = doc(db, "rules_items", ruleId);
      await updateDoc(ruleRef, {
        ...ruleData,
        updatedAt: new Date(),
      });
    } catch (error) {
      console.error("Error updating rule in Firebase:", error);
      throw error;
    }
  }

  // Delete a rule from Firebase
  async deleteRuleFromFirebase(ruleId) {
    try {
      const ruleRef = doc(db, "rules_items", ruleId);
      await deleteDoc(ruleRef);
    } catch (error) {
      console.error("Error deleting rule from Firebase:", error);
      throw error;
    }
  }

  // Get a specific rule from Firebase
  async getRuleFromFirebase(ruleId) {
    try {
      const ruleRef = doc(db, "rules_items", ruleId);
      const ruleDoc = await getDoc(ruleRef);

      if (ruleDoc.exists()) {
        return {
          id: ruleDoc.id,
          ...ruleDoc.data(),
        };
      } else {
        return null;
      }
    } catch (error) {
      console.error("Error fetching rule from Firebase:", error);
      return null;
    }
  }

  // Calculate progressive points
  calculateProgressivePoints(ruleCode, violationCount) {
    const progressiveRules = {
      "PC3.5.B2": {
        // Ngủ trong giờ học
        levels: [
          { level: 1, point: -2 },
          { level: 2, point: -4 },
          { level: 3, point: -6 },
        ],
      },
      "PC5.1.B6": {
        // Đi học trễ
        levels: [
          { level: 1, point: -2 },
          { level: 2, point: -4 },
          { level: 3, point: -6 },
        ],
      },
    };

    const rule = progressiveRules[ruleCode];
    if (!rule) return 0;

    const level = Math.min(violationCount, rule.levels.length);
    return rule.levels[level - 1]?.point || 0;
  }

  // Calculate conditional points
  calculateConditionalPoints(ruleCode, condition) {
    const conditionalRules = {
      "PC3.5.C1": {
        // Tham gia hoặc tổ chức hội thảo, ngày hội nghề nghiệp
        conditions: [
          { type: "tham_gia", point: 2 },
          { type: "to_chuc", point: 4 },
        ],
      },
      "PC3.5.C2": {
        // Đạt giải trong các kỳ thi
        conditions: [
          { type: "quan", point: 2 },
          { type: "thanh_pho", point: 4 },
          { type: "quoc_gia", point: 6 },
        ],
      },
      "PC4.2.C1": {
        // Tham gia hoặc tổ chức, viết bài về gương người tốt
        conditions: [
          { type: "tham_gia", point: 2 },
          { type: "to_chuc", point: 4 },
          { type: "viet_bai", point: 3 },
        ],
      },
      "PC5.2.C1": {
        // Tham gia hoặc tổ chức sự kiện về tiết kiệm
        conditions: [
          { type: "tham_gia", point: 2 },
          { type: "to_chuc", point: 4 },
        ],
      },
      "PC5.4.C1": {
        // Tham gia hoặc tổ chức hoạt động về bổn phận
        conditions: [
          { type: "tham_gia", point: 2 },
          { type: "to_chuc", point: 4 },
        ],
      },
      "PC5.6.B1": {
        // Vắng các buổi lễ
        conditions: [
          { type: "co_ly_do", point: -1 },
          { type: "khong_ly_do", point: -2 },
        ],
      },
    };

    const rule = conditionalRules[ruleCode];
    if (!rule) return 0;

    const matchedCondition = rule.conditions.find((c) => c.type === condition);
    return matchedCondition?.point || 0;
  }

  // Get student behavior violations
  async getStudentViolations(studentId) {
    try {
      const violationsRef = collection(db, "studentViolations");
      const q = query(
        violationsRef,
        where("studentId", "==", studentId),
        orderBy("createdAt", "desc")
      );
      const querySnapshot = await getDocs(q);

      return querySnapshot.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
        createdAt: doc.data().createdAt?.toDate?.() || new Date(),
      }));
    } catch (error) {
      console.error("Error fetching student violations:", error);
      return [];
    }
  }

  // Add student violation
  async addStudentViolation(violationData) {
    try {
      const violationsRef = collection(db, "studentViolations");
      const docRef = await addDoc(violationsRef, {
        ...violationData,
        createdAt: new Date(),
      });
      return docRef.id;
    } catch (error) {
      console.error("Error adding student violation:", error);
      throw error;
    }
  }

  // Get violation count for a specific rule
  async getViolationCount(studentId, ruleCode) {
    try {
      const violationsRef = collection(db, "studentViolations");
      const q = query(
        violationsRef,
        where("studentId", "==", studentId),
        where("ruleCode", "==", ruleCode)
      );
      const querySnapshot = await getDocs(q);
      return querySnapshot.size;
    } catch (error) {
      console.error("Error getting violation count:", error);
      return 0;
    }
  }
}

// Create and export a singleton instance
const rulesService = new RulesService();
export default rulesService;

File: back-end/services/rules_service.py
import json
from typing import Dict, List, Optional, Union
from datetime import datetime

class RulesService:
    def __init__(self):
        # Load JSON data
        with open('competech/admin/ccpad/src/components/bieuhien.json', 'r', encoding='utf-8') as f:
            self.behaviors_data = json.load(f)
        
        with open('competech/admin/ccpad/src/components/diemcong.json', 'r', encoding='utf-8') as f:
            self.plus_data = json.load(f)
        
        with open('competech/admin/ccpad/src/components/diemtru.json', 'r', encoding='utf-8') as f:
            self.minus_data = json.load(f)

    def calculate_total_behavior_points(self) -> int:
        """Calculate total points for behaviors"""
        total = 0
        for category in self.behaviors_data.values():
            for criterion in category['criteria'].values():
                total += criterion['point']
        return total

    def get_next_code(self, parent_code: str, rule_type: str) -> str:
        """Generate next code for a new rule"""
        if rule_type == 'behaviors':
            data = self.behaviors_data
            suffix = 'A'
        elif rule_type == 'plus':
            data = self.plus_data
            suffix = 'C'
        else:  # minus
            data = self.minus_data
            suffix = 'B'

        # Find the parent criterion
        for category in data.values():
            for criterion in category['criteria'].values():
                if criterion['code'] == parent_code:
                    items = criterion.get('behaviors', criterion.get('plus', criterion.get('minus', {})))
                    existing_codes = list(items.keys())
                    
                    if not existing_codes:
                        return f"{parent_code}.{suffix}1"
                    
                    # Find the last code and increment
                    last_code = existing_codes[-1]
                    match = last_code.split('.')[-1]
                    if match.startswith(suffix):
                        try:
                            number = int(match[1:])
                            return f"{parent_code}.{suffix}{number + 1}"
                        except ValueError:
                            pass
                    
                    return f"{parent_code}.{suffix}{len(existing_codes) + 1}"
        
        return f"{parent_code}.{suffix}1"

    def calculate_progressive_points(self, rule_code: str, violation_count: int) -> int:
        """Calculate points for progressive violations"""
        # Find the rule in minus data
        for category in self.minus_data.values():
            for criterion in category['criteria'].values():
                if 'minus' in criterion:
                    for minus_item in criterion['minus'].values():
                        if minus_item['code'] == rule_code and minus_item.get('point') == 'progressive':
                            levels = minus_item.get('levels', [])
                            if levels:
                                # Find the appropriate level based on violation count
                                for level in levels:
                                    if level['level'] == violation_count:
                                        return level['point']
                                # If violation count exceeds max level, use the highest level
                                if levels:
                                    return levels[-1]['point']
        return 0

    def calculate_conditional_points(self, rule_code: str, condition_type: str, context: Dict = None) -> int:
        """Calculate points for conditional violations/rewards"""
        # Find the rule in data
        for data_source in [self.minus_data, self.plus_data]:
            for category in data_source.values():
                for criterion in category['criteria'].values():
                    items = criterion.get('minus', criterion.get('plus', {}))
                    for item in items.values():
                        if item['code'] == rule_code and item.get('point') == 'conditional':
                            conditions = item.get('conditions', [])
                            for condition in conditions:
                                if condition.get('type') == condition_type:
                                    return condition['point']
        return 0

    def add_rule(self, rule_data: Dict) -> Dict:
        """Add a new rule to the appropriate JSON file"""
        rule_type = rule_data.get('type')
        parent_code = rule_data.get('parentCode')
        code = rule_data.get('code')
        description = rule_data.get('description')
        point = rule_data.get('point')
        point_type = rule_data.get('pointType')

        # Validate total points for behaviors
        if rule_type == 'behaviors':
            current_total = self.calculate_total_behavior_points()
            if current_total + point > 100:
                raise ValueError("Tổng điểm biểu hiện không được vượt quá 100 điểm")

        # Prepare the new rule
        new_rule = {
            'code': code,
            'description': description,
            'point': point if point_type == 'fixed' else point_type
        }

        # Add progressive levels if applicable
        if point_type == 'progressive':
            new_rule['levels'] = rule_data.get('levels', [])

        # Add conditions if applicable
        if point_type == 'conditional':
            new_rule['conditions'] = rule_data.get('conditions', [])

        # Add to appropriate data structure
        if rule_type == 'behaviors':
            self._add_to_behaviors(parent_code, new_rule)
        elif rule_type == 'plus':
            self._add_to_plus(parent_code, new_rule)
        else:  # minus
            self._add_to_minus(parent_code, new_rule)

        return {
            'success': True,
            'message': 'Thêm nội quy thành công',
            'rule': new_rule
        }

    def _add_to_behaviors(self, parent_code: str, rule: Dict):
        """Add rule to behaviors data"""
        for category in self.behaviors_data.values():
            for criterion in category['criteria'].values():
                if criterion['code'] == parent_code:
                    if 'behaviors' not in criterion:
                        criterion['behaviors'] = {}
                    criterion['behaviors'][rule['code']] = rule
                    return

    def _add_to_plus(self, parent_code: str, rule: Dict):
        """Add rule to plus data"""
        for category in self.plus_data.values():
            for criterion in category['criteria'].values():
                if criterion['code'] == parent_code:
                    if 'plus' not in criterion:
                        criterion['plus'] = {}
                    criterion['plus'][rule['code']] = rule
                    return

    def _add_to_minus(self, parent_code: str, rule: Dict):
        """Add rule to minus data"""
        for category in self.minus_data.values():
            for criterion in category['criteria'].values():
                if criterion['code'] == parent_code:
                    if 'minus' not in criterion:
                        criterion['minus'] = {}
                    criterion['minus'][rule['code']] = rule
                    return

    def update_rule(self, rule_code: str, rule_data: Dict) -> Dict:
        """Update an existing rule"""
        rule_type = rule_data.get('type')
        
        # Find and update the rule
        if rule_type == 'behaviors':
            success = self._update_in_behaviors(rule_code, rule_data)
        elif rule_type == 'plus':
            success = self._update_in_plus(rule_code, rule_data)
        else:  # minus
            success = self._update_in_minus(rule_code, rule_data)

        if success:
            return {
                'success': True,
                'message': 'Cập nhật nội quy thành công'
            }
        else:
            return {
                'success': False,
                'message': 'Không tìm thấy nội quy cần cập nhật'
            }

    def _update_in_behaviors(self, rule_code: str, rule_data: Dict) -> bool:
        """Update rule in behaviors data"""
        for category in self.behaviors_data.values():
            for criterion in category['criteria'].values():
                if 'behaviors' in criterion and rule_code in criterion['behaviors']:
                    criterion['behaviors'][rule_code].update(rule_data)
                    return True
        return False

    def _update_in_plus(self, rule_code: str, rule_data: Dict) -> bool:
        """Update rule in plus data"""
        for category in self.plus_data.values():
            for criterion in category['criteria'].values():
                if 'plus' in criterion and rule_code in criterion['plus']:
                    criterion['plus'][rule_code].update(rule_data)
                    return True
        return False

    def _update_in_minus(self, rule_code: str, rule_data: Dict) -> bool:
        """Update rule in minus data"""
        for category in self.minus_data.values():
            for criterion in category['criteria'].values():
                if 'minus' in criterion and rule_code in criterion['minus']:
                    criterion['minus'][rule_code].update(rule_data)
                    return True
        return False

    def delete_rule(self, rule_code: str, rule_type: str) -> Dict:
        """Delete a rule"""
        if rule_type == 'behaviors':
            success = self._delete_from_behaviors(rule_code)
        elif rule_type == 'plus':
            success = self._delete_from_plus(rule_code)
        else:  # minus
            success = self._delete_from_minus(rule_code)

        if success:
            return {
                'success': True,
                'message': 'Xóa nội quy thành công'
            }
        else:
            return {
                'success': False,
                'message': 'Không tìm thấy nội quy cần xóa'
            }

    def _delete_from_behaviors(self, rule_code: str) -> bool:
        """Delete rule from behaviors data"""
        for category in self.behaviors_data.values():
            for criterion in category['criteria'].values():
                if 'behaviors' in criterion and rule_code in criterion['behaviors']:
                    del criterion['behaviors'][rule_code]
                    return True
        return False

    def _delete_from_plus(self, rule_code: str) -> bool:
        """Delete rule from plus data"""
        for category in self.plus_data.values():
            for criterion in category['criteria'].values():
                if 'plus' in criterion and rule_code in criterion['plus']:
                    del criterion['plus'][rule_code]
                    return True
        return False

    def _delete_from_minus(self, rule_code: str) -> bool:
        """Delete rule from minus data"""
        for category in self.minus_data.values():
            for criterion in category['criteria'].values():
                if 'minus' in criterion and rule_code in criterion['minus']:
                    del criterion['minus'][rule_code]
                    return True
        return False

    def save_data(self):
        """Save all data back to JSON files"""
        with open('competech/admin/ccpad/src/components/bieuhien.json', 'w', encoding='utf-8') as f:
            json.dump(self.behaviors_data, f, ensure_ascii=False, indent=2)
        
        with open('competech/admin/ccpad/src/components/diemcong.json', 'w', encoding='utf-8') as f:
            json.dump(self.plus_data, f, ensure_ascii=False, indent=2)
        
        with open('competech/admin/ccpad/src/components/diemtru.json', 'w', encoding='utf-8') as f:
            json.dump(self.minus_data, f, ensure_ascii=False, indent=2)

    def get_violation_history(self, student_id: str, rule_code: str) -> int:
        """Get violation count for a specific rule and student"""
        # This would typically query a database
        # For now, return a mock value
        return 0

    def apply_rule(self, student_id: str, rule_code: str, rule_type: str, context: Dict = None) -> Dict:
        """Apply a rule and calculate points"""
        if rule_type == 'behaviors':
            return self._apply_behavior_rule(student_id, rule_code)
        elif rule_type == 'plus':
            return self._apply_plus_rule(student_id, rule_code, context)
        else:  # minus
            return self._apply_minus_rule(student_id, rule_code, context)

    def _apply_behavior_rule(self, student_id: str, rule_code: str) -> Dict:
        """Apply behavior rule"""
        for category in self.behaviors_data.values():
            for criterion in category['criteria'].values():
                if 'behaviors' in criterion and rule_code in criterion['behaviors']:
                    rule = criterion['behaviors'][rule_code]
                    return {
                        'points': rule['point'],
                        'description': rule['description'],
                        'type': 'behavior'
                    }
        return {'points': 0, 'description': 'Không tìm thấy quy tắc', 'type': 'behavior'}

    def _apply_plus_rule(self, student_id: str, rule_code: str, context: Dict = None) -> Dict:
        """Apply plus rule"""
        for category in self.plus_data.values():
            for criterion in category['criteria'].values():
                if 'plus' in criterion and rule_code in criterion['plus']:
                    rule = criterion['plus'][rule_code]
                    if rule.get('point') == 'conditional':
                        # Handle conditional logic
                        points = self.calculate_conditional_points(rule_code, context.get('condition_type', ''), context)
                    else:
                        points = rule['point']
                    
                    return {
                        'points': points,
                        'description': rule['description'],
                        'type': 'plus'
                    }
        return {'points': 0, 'description': 'Không tìm thấy quy tắc', 'type': 'plus'}

    def _apply_minus_rule(self, student_id: str, rule_code: str, context: Dict = None) -> Dict:
        """Apply minus rule"""
        for category in self.minus_data.values():
            for criterion in category['criteria'].values():
                if 'minus' in criterion and rule_code in criterion['minus']:
                    rule = criterion['minus'][rule_code]
                    
                    if rule.get('point') == 'progressive':
                        # Get violation count and calculate progressive points
                        violation_count = self.get_violation_history(student_id, rule_code) + 1
                        points = self.calculate_progressive_points(rule_code, violation_count)
                    elif rule.get('point') == 'conditional':
                        # Handle conditional logic
                        points = self.calculate_conditional_points(rule_code, context.get('condition_type', ''), context)
                    else:
                        points = rule['point']
                    
                    return {
                        'points': points,
                        'description': rule['description'],
                        'type': 'minus',
                        'violation_count': violation_count if rule.get('point') == 'progressive' else None
                    }
        return {'points': 0, 'description': 'Không tìm thấy quy tắc', 'type': 'minus'}

